risc0/circuit/rv32im/src/execute/segment.rs

```diff
@@ -46,3 +54,48 @@ pub struct Segment {
 
     pub index: u64,
 }
+
+impl Segment {
+    pub fn encode(&self) -> Result<Vec<u8>> {
+        Ok(postcard::to_stdvec(self)?)
+    }
+
+    pub fn decode(bytes: &[u8]) -> Result<Self> {
+        Ok(postcard::from_bytes(bytes)?)
+    }
+
+    pub fn execute(&self) -> Result<()> {
+        let handler = SegmentSyscallHandler {
+            segment: self,
+            read_pos: AtomicUsize::new(0),
+            write_pos: AtomicUsize::new(0),
+        };
+        Executor::new(self.partial_image.clone(), &handler, None, vec![]).run(
+            self.po2 as usize,
+            MAX_INSN_CYCLES,
+            testutil::DEFAULT_SESSION_LIMIT,
+            |_| Ok(()),
+        )?;
+        Ok(())
+    }
+}
+
+struct SegmentSyscallHandler<'a> {
+    segment: &'a Segment,
+    read_pos: AtomicUsize,
+    write_pos: AtomicUsize,
```

Comment:
Guess this could have been `Cell<usize>` I think, but this is fine

Reply:
I couldn't figure out how to do `borrow_mut` without `self` also being `mut`.